### ISSUE TITLE
+default in PodSpec has too much ripple

### ISSUE THREAD

**Issue by thockin** [kubernetes/kubernetes#120838]
@alexzielenski @apelisse 

It's well understood that PodSpec is embedded in all sorts of types, so any new defaults in there change all those other types.  Usually we don't want that, and so the hand-coded defaulting has ` SetDefaults_PodSpec()` (which captures the existing ones) and `SetDefaults_Pod()` (which only triggers on actual Pod, which is where most new defaults should go).

There's no way to denote this with `+default`.  Maybe we don't want to because it is so specific to this case, but I wanted to file it to think about.

**Comment by apelisse** [comment]
Yeah, good point, Alex mentioned that to me already. Let's see what we can do ...

**Comment by apelisse** [comment]
Would it have made sense to use different types? I don't think we can change that now though. On the other hand, we've done a lot of "copying types" in kube-openapi to improve the serialization speed, and there are a lot of compile-time checks when you copy a type into another, so it's not entirely impossible to do now I would say.

**Comment by thockin** [comment]
I think we could have done (or could do) something to explicitly say "this is the same structure but different rules" in a bunch of places.  That said, it's not clear to me that just having two types would always work.

Sometimes we should have simply done deep-copy of the type definition or even deep-inlining.  I would love to see us spend some time disentangling this.  It's really a maintenance nightmare and a long term drag on our ability to do new things in a safe way.  Like it or not, PodSpec is going to continue to get new fields :)

**Comment by thockin** [comment]
Since ##124220 merged we WILL be able to set pod defaults...in a few more releases (once the oldest kubelets are all updated to 1.30)
